File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root,Integer.MIN_VALUE,Integer.MAX_VALUE);
        
    }
    public boolean helper(TreeNode root, long low, long high){
        if(root==null) return true;
        
        if(low>root.val || root.val>high)
            return false;        
        
        return helper(root.right,root.val+1L,high) &&
            helper(root.left,low,root.val-1L);
            
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def validate(self,root,min,max):
        if not root:
            return True
        # if root.val<min or root.val>max:
        #     return False
        return min<root.val<max and self.validate(root.left,min,root.val) and self.validate(root.right,root.val,max)
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.validate(root,float('-inf'),float('inf'))
        